CppImport/src/macro/MacroExpansions.cpp

```diff
@@ -0,0 +1,244 @@
+{
+	expansionCache_.clear();
+	expansions_.clear();
+}
+
+QVector<MacroExpansion*> MacroExpansions::topLevelExpansions() const
+{
+	QVector<MacroExpansion*> result;
+	for (auto expansion : expansions_)
+		if (!expansion->parent())
+			result.append(expansion);
+
+	return result;
+}
+
+
+MacroExpansion* MacroExpansions::expansion(clang::SourceLocation loc) const
+{
+	MacroExpansion* expansion = immediateExpansion(loc);
+	MacroExpansion* last = expansion;
+
+	if (expansion)
+	{
+		do
+		{
+			last = expansion;
+			loc = clang_.sourceManager()->getImmediateExpansionRange(loc).first;
+			expansion = immediateExpansion(loc);
+		} while (expansion && expansion->isChildOf(last));
+	}
+
+	return last;
+}
+
+MacroExpansion* MacroExpansions::immediateExpansion(clang::SourceLocation loc) const
+{
+	auto expansion = clang_.immediateMacroLocation(loc);
+	for (auto i = 0; i < expansions_.size(); i++)
+		if (expansions_[i]->range().getBegin() == expansion) return expansions_[i];
+
+	/*
+	 * if we have not found an immediate expansion for loc then we try again using the found immediateExpansionLoc.
+	 * this can happen in case of token concatenation or stringification where the first expansion location would point
+	 * to the location of the concatenated token or stringification result.
+	 */
+	expansion = clang_.immediateMacroLocation(expansion);
+	for (auto i = 0; i < expansions_.size(); i++)
+		if (expansions_[i]->range().getBegin() == expansion) return expansions_[i];
+
+	return nullptr;
+}
+
+
+QSet<MacroExpansion*> MacroExpansions::expansions(Model::Node* node)
+{
+	Q_ASSERT(node);
+
+	auto it = expansionCache_.find(node);
+	if (it != expansionCache_.end()) return *it;
+
+	QSet<MacroExpansion*> result;
+	if (auto n = envisionToClangMap_.closestParentWithAstMapping(node))
+		if (envisionToClangMap_.contains(n))
+			for (auto range : envisionToClangMap_.get(n))
+				if (auto exp = expansion(range.getBegin()))
+					result.insert(exp);
+
+	expansionCache_.insert(node, result);
+	return result;
+}
+
+
+QVector<Model::Node*> MacroExpansions::topLevelNodes(MacroExpansion* expansion, NodeOriginType filter)
+{
+	Q_ASSERT(expansion);
+
+	QVector<Model::Node*> allTLExpansionNodes;
+	for (auto node : envisionToClangMap_.nodes())
+		for (auto range : envisionToClangMap_.get(node))
+		{
+			bool isCandidate = filter == Transitive ?
+						// return true if the node ultimately expands to the source location of the expansion either by
+						// being directly generated by expansion or by one of its child expansions.
+						clang_.sourceManager()->getExpansionLoc(range.getBegin()) == expansion->range().getBegin() :
+						// only return true if the node is actually generated in expansion.
+						expansions(node).contains(expansion);
+
+			if (isCandidate)
+			{
+				allTLExpansionNodes.append(node);
+				break;
+			}
+		}
+
+	QVector<Model::Node*> result = NodeHelpers::topLevelNodes(allTLExpansionNodes);
+
+	qSort(result.begin(), result.end(),
```

Comment:
One way to change this function is this:
- compute the lowest common ancestor for `e1` and `e2`, let's call this `common`
- Note that `common` might be null
- In addition to the common ancestor, also compute the last node before the `common` ancestor on the path to that ancestor. So compute the `p1` and `p2`, in this chain [`e1`, `e1->parent()`, ..., `p1`, `p1->parent()` == `common`], [`e2`, `e2->parent()`, ..., `p2`, `p2->parent()` == `common`]. Note that `p1 != p2`.
- If `common` is null, you can return `p1 < p2` (not `e1` < `e2`).
- Otherwise, you need an ordering based on `common`. There should be only two cases:
- If `common` is a list do what you're doing currently.
- If `common` is `CompositeNode` then

    auto index1 = common->indexOf(p1);
    auto index2 = common->indexOf(p2);
    Q_ASSERT(index1.isValid() && index2.isValid())
    return index1.level() < index2.level() || (index1.level()==index2.level() && index1.index() < index2.index())
- There should be only these 3 options for `common`, so use `Q_ASSERT(false)` to mark any other options and we'll investigate if we hit those.